url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageARMv5.java

```diff
@@ -0,0 +1,4246 @@
+        int vaddr, paddr;
+
+        if (!exec) {
+            printDisasm(inst,
+                    String.format("strh%s", inst.getCondFieldName()),
+                    String.format("%s, %s", getRegName(rd),
+                            getAddrMode3Name(inst)));
+            return;
+        }
+
+        if (!inst.satisfiesCond(getCPSR())) {
+            return;
+        }
+
+        if (p) {
+            //プリインデクス
+            vaddr = offset;
+        } else {
+            //ポストインデクス
+            vaddr = getReg(rn);
+        }
+
+        paddr = getMMU().translate(vaddr, 2, false, getCPSR().isPrivMode(), false);
+        if (getMMU().isFault()) {
+            getMMU().clearFault();
+            return;
+        }
+
+        if (!tryWrite_a32(paddr, 2)) {
+            raiseException(ARMv5.EXCEPT_ABT_DATA,
+                    String.format("strh [%08x]", paddr));
+            return;
+        }
+        write16_a32(paddr, (short) getReg(rd));
+
+        if (!p || w) {
+            //ベースレジスタを更新する
+            //条件は !(p && !w) と等価、つまり P, W ビットが
+            //オフセットアドレス以外の指定なら Rn を書き換える
+            setReg(rn, offset);
+        }
+    }
+
+    /**
+     * レジスタダブルワードストア命令。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeStrd(InstructionARM inst, boolean exec) {
+        boolean p = inst.getBit(24);
+        boolean w = inst.getBit(21);
+        int rn = inst.getRnField();
+        int rd = inst.getRdField();
+        int offset = getAddrMode3(inst);
+        int vaddr, paddr;
+
+        if (!exec) {
+            printDisasm(inst,
+                    String.format("strd%s", inst.getCondFieldName()),
+                    String.format("%s, %s", getRegName(rd),
+                            getAddrMode3Name(inst)));
+            return;
+        }
+
+        if (!inst.satisfiesCond(getCPSR())) {
+            return;
+        }
+
+        if (p) {
+            //プリインデクス
+            vaddr = offset;
+        } else {
+            //ポストインデクス
+            vaddr = getReg(rn);
+        }
+
+        paddr = getMMU().translate(vaddr, 4, false, getCPSR().isPrivMode(), false);
+        if (getMMU().isFault()) {
+            getMMU().clearFault();
+            return;
+        }
+
+        if (!tryWrite_a32(paddr, 4) || !tryWrite_a32(paddr + 4, 4)) {
+            raiseException(ARMv5.EXCEPT_ABT_DATA,
+                    String.format("strd [%08x]", paddr));
+            return;
+        }
+        write32_a32(paddr, getReg(rd));
+        write32_a32(paddr + 4, getReg(rd + 1));
+
+        if (!p || w) {
+            //ベースレジスタを更新する
+            //条件は !(p && !w) と等価、つまり P, W ビットが
+            //オフセットアドレス以外の指定なら Rn を書き換える
+            setReg(rn, offset);
+        }
+    }
+
+    /**
+     * ロードマルチプル命令。
+     *
+     * @param inst ARM 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executeLdm1(InstructionARM inst, boolean exec) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 3388-3407 into a separate method.